lib/graphql/static_validation/rules/fields_are_defined_on_type.rb

```diff
@@ -16,7 +15,7 @@ def validate(context)
       private
 
       def validate_field(context, ast_field, parent_type, parent)
-        if parent_type.kind.union?
+        if parent_type.kind.union? && ast_field.name != '__typename'
```

Comment:
As per the spec:

> Because unions do not define fields, fields may not be directly selected from a union‐typed selection set, with the exception of the meta‐field `__typename`.